### PR TITLE
refactor!: unwrap BasicBlock enum

### PR DESCRIPTION
**Pull request by ss2165**
BREAKING_CHANGES: DFB and Exit are now standalone structs. A level of nesting in serialization is also removed for these operations.

### CODE REVIEW

File: src/ops/validate.rs [hunk 1]
```diff
@@ -309,20 +305,26 @@ impl ValidateOp for BasicBlock {
...
+    }
+}
+
+impl ValidateOp for Exit {
+    /// Returns the set of allowed parent operation types.
+    fn validity_flags(&self) -> OpValidityFlags {
+        Default::default()
+    }
+
+    /// Validate the ordered list of children.
```
Comment (acl-cqc) [inline src/ops/validate.rs:322]: This is the default impl, so can skip

File: src/ops/validate.rs [hunk 2]
```diff
@@ -309,20 +305,26 @@ impl ValidateOp for BasicBlock {
...
+        let node_outputs: TypeRow = [&[tuple_sum_type], self.other_outputs.as_ref()]
+            .concat()
+            .into();
+        validate_io_nodes(&self.inputs, &node_outputs, "basic block graph", children)
+    }
+}
+
+impl ValidateOp for Exit {
+    /// Returns the set of allowed parent operation types.
+    fn validity_flags(&self) -> OpValidityFlags {
```
Comment (acl-cqc) [inline src/ops/validate.rs:318]: This is *also* the default impl, so can skip

**Comment by acl-cqc** [comment]
Ok, so the cunning followup which might make this `DataflowParent` trait (currently unused) really fly, is to provide an
```
impl<T: DataflowParent> ValidateOp for T {...}
```

noting that all impls (for such T) have the same `validity_flags` and I *think* that the differences in `validate_op_children` are captured exacttly by `inner_signature`....

**Comment by acl-cqc** [comment]
My other thought is nomenclature. How about `DFB` -> `BasicBlock` and `Exit` -> `ExitBlock`? (DFB is very similar to DFG, I admit, but the one character that's different doesn't quite seem right)

**Comment by ss2165** [comment]
Yeah those new names seem a bit nicer - but that's definitely a broader spec-inclusive issue

File: src/ops/controlflow.rs [hunk 3]
```diff
@@ -190,43 +180,73 @@ impl OpTrait for BasicBlock {
     }
 
     fn dataflow_signature(&self) -> Option<FunctionType> {
-        Some(match self {
-            BasicBlock::DFB {
-                extension_delta, ..
-            } => FunctionType::new(type_row![], type_row![]).with_extension_delta(extension_delta),
-            BasicBlock::Exit { .. } => FunctionType::new(type_row![], type_row![]),
-        })
+        Some(
```
Comment (acl-cqc) [inline src/ops/controlflow.rs:183]: Hmmm, I highly suspect this should be `None` (but instead we should implement `fn extension_delta` which by default delegates to this). However, that's not really part of this PR

File: src/ops/validate.rs [hunk 4]
```diff
@@ -345,7 +263,7 @@ impl ValidateOp for super::Case {
         children: impl DoubleEndedIterator<Item = (NodeIndex, &'a OpType)>,
     ) -> Result<(), ChildrenValidationError> {
         let sig = self.inner_signature();
-        validate_io_nodes(&sig.input, &sig.output, "Conditional", children)
+        validate_io_nodes(&sig.input, &sig.output, "DataflowParent", children)
```
Comment (acl-cqc) [inline src/ops/validate.rs:266]: I guess there is no handy way to get a name? "DataflowParent" seems fine if nothing springs to mind

Comment (ss2165) [reply src/ops/validate.rs:266]: we can add a dependency on the `OpName` trait and use that but this function
(and everything subsequently including errors) expects static strings so its
just a big annoying change to make.

Comment (ss2165) [reply src/ops/validate.rs]: have gone with `std::any::type_name` instead

File: src/ops/controlflow.rs [hunk 5]
```diff
@@ -190,43 +180,73 @@ impl OpTrait for BasicBlock {
...
+        match dir {
+            Direction::Incoming => 1,
+            Direction::Outgoing => 0,
         }
     }
+}
 
+/// Functionality shared by DFB and Exit CFG block types.
+pub trait BasicBlock {
+    /// The input signature of the contained dataflow graph.
```
Comment (acl-cqc) [inline src/ops/controlflow.rs:227]: The dataflow signature of the values input to the block? (Exit blocks do not have a "contained dataflow graph")

**Comment by acl-cqc** [comment]
Ok, so - on the naming front: the spec says DFB and Exit are the two types of Basic Block; this PR keeps DFB and Exit but largely drops the notion of "Basic Block"! So I think it would be fine to do some name changes in the code (with appropriate references to the spec - `struct DataflowBlock` is (in the spec) the dataflow subtype of Basic Block, and `struct ExitBlock` is (in the spec) the exit subtype of Basic Block).

But, generally, I'm happy to fold this into #754 and then follow up with a renaming PR. (The *ideal* would probably be to unwrap basicblock and rename in a first PR and then do #754 after, but I'm not going to make you do that!)

**Comment by acl-cqc** [comment]
(I think we should not put #754 in without this, because of the rather dubious inner-signature for exit blocks)

File: src/ops/validate.rs (hunk 4 above)
Comment (acl-cqc) [reply src/ops/validate.rs]: using `T`! Nice :-)

**Comment by ss2165** [comment]
I've taken your suggestion and reordered the PRs, this one is being closed in favour of #781
